### ISSUE TITLE
View update propagation should apply backpressure to coordinator

### ISSUE THREAD

**Issue by duarten** [scylladb/scylla#2538]
Currently we propagate view updates asynchronously from a base replica to its paired view replicas.

The asynchronous propagation is meant to lower latency and prevent the base replica availability from being impacted by a paired view replica's availability.

In case a view replica is down, a view update is persisted in the batchlog, to be retried later.

During normal operation, a view replica may be overwhelmed and need to apply backpressure to the base replica, which needs to queue the view update.

To avoid unbounded queue growth, we can propagate this backpressure to the client by delaying normal base table writes, taking into account the queue size.

A simpler approach, though, is to make view propagation synchronous. A base table write will have higher latency, the increase being determined by the slowest view replica. Assuming there's a mechanism ensuring view updates will eventually reach the view replica, we can skip writing it to that replica's commitlog.

The issue becomes what to do if a view replica is down. In this case we can still rely on the batchlog to remember the view update, by writing view updates to it after a failed attempt to reach a view replica (or by using failure detector state), but before locally recording the base update.

A more audacious approach would be to do away with the bachtlog write. If the base write is being done with CF < RF, then we can attempt to send the update to another view replica, paired to other base replicas. We could choose those view replicas randomly. In case a write is being done with CF = RF, then we can just omit the write and rely on repair to eventually heal the view replica.

This repair could be normal repair between view replicas, repairing a view replica with its paired base replica, or probabilistic read repair between base and view (note that only the first form of repair is currentely implemented).

Finally, note that synchronously propagating view updates helps with view building, which could then reuse code that currently doesn't wait for the view replica's reply. Waiting for the reply is needed to locally update view building status.

**Comment by duarten** [comment]
cc @avikivity @nyh

**Comment by avikivity** [comment]
I'm in favor of synchronous updates, but picking a random replica is dangerous, no? Suppose we have two base replicas failing to update their view replicas, and picking the same alternate view replica to update. Then we will have updated just one view replica, below our consistency level.

Perhaps the base replica write should return some information to the caller, so it can decide on whether CL was achieved or not.

**Comment by slivne** [comment]
I am against dropping asynchronous propagation - we can have multiple views
to a single base - will we wait for all the replica's to return at all
times (even if we have a very low write rate ?)

> During normal operation, a view replica may be overwhelmed and need to
apply backpressure to the base replica, which needs to queue the view
update.

> To avoid unbounded queue growth, we can propagate this backpressure to
the client by delaying normal base table writes, taking into account the
queue size.

Why not use a similar back pressure mechanism as we have for standard
writes - e.g. move the background replica writes to become foreground
replica writes based on memory/queue length.

**Comment by avikivity** [comment]
> I am against dropping asynchronous propagation - we can have multiple views
to a single base - will we wait for all the replica's to return at all
times (even if we have a very low write rate ?)

Why not? There is the latency of the extra hop, but it's still very low. Multiple view writes can proceed in parallel.

> Why not use a similar back pressure mechanism as we have for standard
writes - e.g. move the background replica writes to become foreground
replica writes based on memory/queue length.

Because this is done on the replica side.

**Comment by slivne** [comment]
On Sun, Jul 2, 2017 at 5:10 PM, Avi Kivity <notifications@github.com> wrote:

> I am against dropping asynchronous propagation - we can have multiple views
> to a single base - will we wait for all the replica's to return at all
> times (even if we have a very low write rate ?)
>
> Why not? There is the latency of the extra hop, but it's still very low.
> Multiple view writes can proceed in parallel.
>
The same is true with regular write that have CL < RF , we incur additional
latency - its small but exists - we still return faster and have background
writes.

**Comment by avikivity** [comment]
> The same is true with regular write that have CL < RF , we incur additional
latency - its small but exists - we still return faster and have background
writes.

It's not the same, we can survive a failure with CL<ALL.

If we had a really fast failure detector, then it would be better to always write synchronously within the same data center and return when all writes succeeded, or when the failure detector announced a failure. The small gain in latency is not worth the complexity.

**Comment by gleb-cloudius** [comment]
On Sun, Jul 02, 2017 at 02:49:48PM +0000, Avi Kivity wrote:
> > The same is true with regular write that have CL < RF , we incur additional
> latency - its small but exists - we still return faster and have background
> writes.
> 
> It's not the same, we can survive a failure with CL<ALL.
> 
> If we had a really fast failure detector, then it would be better to always write synchronously within the same data center and return when all writes succeeded, or when the failure detector announced a failure. The small gain in latency is not worth the complexity.
> 
This will slow your writes to a speed of slowest node. If a node is
permanently slower and write load is constant this may be not a bad
things, but in other cases it is.

--
			Gleb.

**Comment by avikivity** [comment]
If your workload is throughput bound (you are pushing as much work as it will bear) then this is not an issue.

If your workload is independent of the cluster (number of reqs/sec not dependent on cluster response), then this is not an issue.

If your workload is latency bound (bounded concurrency), this can be an issue, but the solution is to remove the bound.

Consistency level is not a "make it fast" button. It's "increase availability at the expense of consistency" button.

(for reads it does lower the cluster utilization, so it is somewhat a "make it fast" button)

**Comment by gleb-cloudius** [comment]
On Sun, Jul 02, 2017 at 03:24:29PM +0000, Avi Kivity wrote:
> If your workload is throughput bound (you are pushing as much work as it will bear) then this is not an issue.
> 
Then why wait for local nodes only? Remote DC will slow you down
eventually if it cannot keep up.

> If your workload is independent of the cluster (number of reqs/sec not dependent on cluster response), then this is not an issue.
> 
> If your workload is latency bound (bounded concurrency), this can be an issue, but the solution is to remove the bound.
> 
Removing what bound?

> Consistency level is not a "make it fast" button. It's "increase availability at the expense of consistency" button.
> 
It is used like one though. And in a way it is. It batches writes for
slower nodes (or nodes with bigger latency).

--
			Gleb.

**Comment by nyh** [comment]
I'm a latecomer to this long thread, but please note that in Cassandra, view table writes happen asynchronously - so that alone is good reason to keep this behavior unless we have good reason to change it. There seem to be good reasons for both choices - do we have any indication that Cassandra chose based on wrong information, or on information that is no longer relevant on Scylla?
Also, what's the benefit of moving to synchrnous view update? It would allow us to make a new promise (that  the view is already updated as soon as you get a successful write), but since Cassandra never made this promise, people couldn't rely on....

As I see it, one of the biggest problems with the "synchronous" update is destroying high-availability: Consider a cluster of 3 nodes, RF=3 and we read and write with CL=2 (quorum). The loss of any one of the nodes is tolerated. At first glance, one might think that the same applies to the view table as well: We can also write two replicas there even if one node is down. But that is NOT how the current algorithm, which we copied from Cassandra, works... In the current algorithm, we have a pairing between base and view replicas. So once we chose two base replicas (the two which are up), this defines the two view replicas - and one of those may be down... Yes, we can probably find a way to handle this case too - Duarte mentioned a couple above - perhaps falling back to using the batchlog, or to send the update to an alive but "wrong" paired replica. I particularly like the last idea, but we need to think more carefully about its correctness properties.

**Comment by avikivity** [comment]
>> If your workload is throughput bound (you are pushing as much work as it will bear) then this is not an issue.

>Then why wait for local nodes only? Remote DC will slow you down
eventually if it cannot keep up.

Because it's a different class or latency. If the latency changes from 0.5ms to 1ms, you need double the concurrency. If the latency changes from 0.5ms to 50ms, you need 100X the concurrency.

> Removing what bound?

Concurrency.

>>  Consistency level is not a "make it fast" button. It's "increase availability at the expense of consistency" button.

> It is used like one though. And in a way it is. It batches writes for
slower nodes (or nodes with bigger latency).

At high load (with enough concurrency) writes are batched anyway.

Low CL amplifies the client concurrency and in this way increased efficiency. But the answer to "my write throughput is too low" should not be "reduce your consistency level" (at least within a data center).

**Comment by avikivity** [comment]
@nyh in a throughput bound load (client sends writes as fast as it can) asynchronous work will pile up, and eventually we will run out of memory. There are several options available:

 - do the work synchronously
 - do the work asynchronously, but when a certain amount of backlog is reached, switch to synchronous mode (this is what we do for base table writes with CL<ALL)
 - do the work asynchronously, but when a certain amount of backlog is reached, reject further writes (the downside here is that the client has no idea when it can retry)
 - do the work asynchronously, but when a certain amount of backlog is reached, drop the writes, and rely on repair to fix them

All of the options have costs, but doing things synchronously is both cheapest and most robust, if it doesn't violate any contracts.

**Comment by nyh** [comment]
@avikivity it doesn't violate any contracts, but it increases latency in low-load situations where latencies are the only thing that matters (we have some clients like that). A Cassandra user might be used to view updates being done in the background and doesn't want it to increased latency. But I agree this is just one of many types of users.

**Comment by avikivity** [comment]
The latencies are going to be completely different anyway (and likely lower).

It's still the same class of latency (O(1) hop count within the local data center).

**Comment by avikivity** [comment]
It's going to be different, though, if we do a RMW.

**Comment by slivne** [comment]
The overhead of propogating info to the replica can be much higher

You can have a delete and then an insert

You can have multiple deletes and inserts

**Comment by duarten** [comment]
@avikivity 

> I'm in favor of synchronous updates, but picking a random replica is dangerous, no? Suppose we have two base replicas failing to update their view replicas, and picking the same alternate view replica to update. Then we will have updated just one view replica, below our consistency level.

There isn't any hard guarantee about the consistency levels when it comes to views. It can very well be that for a CL ALL write, none of the view replicas are available. The only guarantee is that the update will eventually propagate to the view replicas. The way I was thinking about this is that, eventually, repair will ensure consistency of each view replica, and everything we do until then is just helping a bit.

> Perhaps the base replica write should return some information to the caller, so it can decide on whether CL was achieved or not.

That's one option, but then we would be adding one extra hope to a synchronous path, which I have a harder time justifying.

Another option is to just send the update to the other view replicas, asynchronously. This has the downside that one of the replicas processes the same update twice.

**Comment by duarten** [comment]
After some discussion, it seems the issue of availability is hard to solve (reminder: the problem is having the view replica's availability impact the base replica's). 

A synchronous request to a failed view replica, that hasn't been detected as such yet, will eventually timeout. As a consequence, the coordinator will most likely timeout the request to the base replica as well, not to mention client-side timeouts. The problem is aggravated by the fact that there can be plenty of concurrent requests timing out.

We could think about solving this by moving the view replica update to the coordinator, in which case the base replicas would just reply with the view updates. The coordinator would then have the ability to abort the view update when some deadline is approaching, while still generating a successful response to the client. An issue with this approach is calculating exactly what that cutoff time should be, which would lead us to forget the ongoing view replica request. (Other issues are higher network usage and higher coordinator side memory usage, which could potentially be solved with variations on this idea.)

For now we'll keep the asynchronous updates and deal with the backpressure issue explicitly.

**Comment by duarten** [comment]
Some further comments on the subject.

For a table with V materialized views, an admitted write at a base replica will result, on average, in V * 2 view updates being generated. These updates are then propagated to a set of view replicas, which grows with the amount of view updates generated. We should hold onto an update until the corresponding view replica has acknowledged it, or until resource constraints force the base replica to drop it.

There is thus a queue[1] of pending and in-flight view updates at the base replica, or, more precisely, a set of queues per paired view replica. At each of those, there's a queue for incoming writes. A process takes view updates from these queues and pushes them to the corresponding node. There should be back-pressure from the node to that process, which adaptively sends more or less view updates (using a TCP Reno approach or similar). This should be implemented by hinted handoff, and if it is not, it's a separate issue.

Since the queues at the base replica grow in proportion to the number of admitted writes, we need to apply back-pressure to them as well. We could use a `backlog_controller` that monitors the queue sizes and affects the shares for the `statement_scheduling_group`. Ideally, there would be one such group per table, so that we can restrict writes to those with pending view updates but not others.

[1] This is #3383, which is kind of intertwined with this issue.

**Comment by slivne** [comment]
For regular writes we move background writes to become foreground writes -
and thus delay the response to the client - providing some backpressure
toward the client.

In MV - AFAIK we do not do that - should we consider moving the base -->
view writes from being background to foreground ?

**Comment by duarten** [comment]
Yes, I didn't mention the part where the queues need a limit; I assume that to be solved by hinted handoff.

Good point on the separate verbs. I don't think this is strictly needed if we apply controls per table; only selected base tables would be throttled. 

Yeah, a controller may not be the best choice. What we want to do is slow down the user writes for a particular table, so we need to park them for a period proportional to the queue size of pending view updates, and eventually shed load.

We must also distinguish between a slow replica and a dead replica (failure detector caveats apply here).

Having foreground base -> view writes was the initial idea we explored, which unfortunately affects base write availability. Slowing down writes while the view update queue is growing has a similar effect to foreground writes, but still allows progress if a view replica is down.

**Comment by gleb-cloudius** [comment]
On Wed, Apr 25, 2018 at 11:21:02PM -0700, Avi Kivity wrote:
> To apply the limit, we need separate verbs for view mutations and base mutations, so we can stop admitting base mutations but still admit view mutations (this allowing other nodes to clear their queue, and also prevent livelocks).
> 
How having different verbs will allow you to admit one and not the
other? It would be possible only if both verbs are running on dedicated
connections. Why not apply the limit on a coordinator?

--
			Gleb.
